packages/react/src/components/Dialog/index.js

```diff
@@ -1,153 +1,231 @@
-      event.stopPropagation();
-      savedOnDismiss();
-    }
-  }
+    const handleClose = (ev) => {
+      if (onClose) {
+        onClose(ev);
+      }
+    };
+
+    const handleBackdropClick = (ev) => {
+      if (ev.target === localRef?.current) {
+        setLocalOpen(false);
+      }
+    };
 
-  useEffect(() => {
-    const changes = hide(document.body, dialogRef.current);
-    return () => {
-      show(changes);
+    const handleTransitionEnd = (ev) => {
+      if (ev.target === localRef?.current) {
+        if (isClosing) {
+          setIsOpen(false);
+          setIsClosing(false);
+          localRef?.current?.close();
+        }
+      }
     };
-  }, []);
-
-  return (
-    <FocusScope
-      {...rest}
-      aria-labelledby={labelledBy}
-      aria-modal="true"
-      initialFocusRef={dialogRef}
-      onKeyDown={onKeyDown}
-      ref={ref}
-      role="dialog"
-      tabIndex="-1">
-      {children}
-    </FocusScope>
-  );
-});
+
+    useEffect(() => {
+      // this progresses closing when isClosing
+      if (isOpen && isClosing) {
+        if (prefersReducedMotion) {
+          setIsOpen(false);
+          setIsClosing(false);
+        }
+        // note prefersReducedMotion case handled by handleTransitionEnd
+      }
+    }, [isClosing, isOpen, prefersReducedMotion]);
+
+    useEffect(() => {
+      if (isOpening) {
+        // dialog should already be displayed set isOpen animates opening
+        setIsOpen(true);
+        setIsOpening(false);
+      }
+    }, [isOpening]);
+
+    useEffect(() => {
+      // this handles localOpen change
+      if (localOpen !== isOpen) {
+        if (localOpen) {
+          if (modal) {
+            localRef.current.showModal();
+          } else {
+            localRef.current.open = true;
+          }
+
+          if (prefersReducedMotion) {
+            setIsOpening(true);
+          } else {
+            /* opening and closing is used in as allow-discrete is not currently supported wide enough
+             * https://caniuse.com/mdn-css_properties_display_is_transitionable */
+
+            // delaying setIsOpening to allow animation
+            window.requestAnimationFrame(() => setIsOpening(true));
+          }
+        } else {
+          // isClosing triggers actual close sequence (allowing for animation)
+          setIsClosing(true);
+        }
+      }
+    }, [isOpen, localOpen, localRef, modal, prefersReducedMotion]);
+
+    useEffect(() => {
+      // triggers open or close sequence
+      setLocalOpen(open);
+    }, [open]);
+
+    return (
+      // eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-noninteractive-element-interactions
+      <dialog
+        {...rest}
+        className={cx(`${prefix}--dialog`)}
+        data-closing={isClosing ? '' : null}
+        data-is-open={localOpen && isOpen ? '' : null}
+        data-width={width}
+        data-modal={modal ? '' : null}
+        data-kind={kind ? kind : null}
```

Comment:
I'm not sure about this. Use of `:where` to bring specificity of carbon styles to `0` is an interesting idea, but I'm not sure about the data attribute approach. It would fragment the css portion of our style guide. Refactoring to take this approach everywhere would be a significant breaking change.

Reply:
I don't disagree, but was suffering a bit of modifier overload with BEM. I still like the B & E but think some class attributes can look very overloaded with BE BEM1 BEM2 BEM 3BEM4. There are a few places where breaks from BEM e.g. `is-visible` which looks much more readable BE M1 M2 M3 M4. 

Presumably you'd like me to change to `B__E--M`